### ISSUE TITLE
Blank node labels/identifiers

### ISSUE THREAD

**Issue by rubensworks** [w3c/sparql-query#66]
As raised by @afs in https://github.com/w3c/sparql-query/pull/65:

> Blank node labels are about the syntax - RDF Concepts (form 1.1) now calls these syntax labels "Blank Node Identifiers" #dfn-blank-node-identifier. So first issue - do we replace all mention of labels with identifier, and mention in 1.2.4 Terminology that label is a common, informal alternative or say somewhere (1.2.4 Terminology) "Blank Node labels" are "Blank Node Identifiers" and continue using labels. Either way round is tricky.

So either we align these the bnode labels and bnode identifiers concepts, or we consistently use one (e.g. bnode identifiers).

**Comment by hartig** [comment]
I am in favor of consistently using one, which then should be "blank node identifier" because that's the term introduced in RDF Concepts. In fact, I would not even mention labels at all in the SPARQL doc; i.e., not even saying that it "is a common, informal alternative". Something like this may instead be mentioned in RDF Concepts. I have created an issue in the RDF Concepts repo to consider this option (https://github.com/w3c/rdf-concepts/issues/38).

**Comment by gkellogg** [comment]
We should strive for consistency in naming, and “blank node identifier” is the term.  It, we should do something to help clarify the persistent confusion between labels and identifiers.

**Comment by afs** [comment]
See assessment : https://github.com/w3c/rdf-concepts/issues/38#issuecomment-1537095857

TL;DR:
SPARQL Query uses both.
Easy enough to consistently use "blank node identifier". The two phrases mean the same thing.

The only thing to be careful of in our documents would be any unqualified use of "label" and "identifier" when the context isn't clear.

I'd leave the grammar rule `BLANK_NODE_LABEL`. It's only in the grammar. It is the specific choice of blank node identifier for the SPARQL language (and Turtle, NT, NQ but not RDF/XML). It is used elsewhere in related languages like SHACL and ShEx.

**Comment by afs** [comment]
Anything more to do after #65 and #69 or can we close this issue on sparql-query?

**Comment by gkellogg** [comment]
There's some text in [4.1.4 Syntax for Blank Node](https://w3c.github.io/sparql-query/spec/#QSynBlankNodes) on `[]`that says it allocates a unique blank node identifier:

> allocate a unique blank node identifier (here "b57") ...

Does the spec require that it allocate a unique identifier? Or, just a unique blank node? On Collections, `()` allocates blank nodes, but not identifiers. It would only be if re-serialized that a blank node identifier would be significant. In generally, allocating blank node identifiers is hard, unless all blank nodes are systematically renamed to avoid collisions.

There's a sentence later on: "generate blank nodes with identifier _:v1 and _:v2 in the results graph", where the language is subtly different than how blank node identifiers are described elsewhere.

**Comment by afs** [comment]
Thanks for catching that - no, ids aren't allocated (I suppose some parsers might implement that way).

Re-serialiation isn't a factor - there are no blank node identifiers after the parser has finished.

I'll do a PR to fix 4.1.4.

**Comment by rubensworks** [comment]
It looks like this issue has been fully resolved now.

(feel free to re-open if anyone disagrees)
